File: c#/EmailMarketingApi/AppCode/EmailMkt.cs
/*
 *  Copyright (c) 2009, Locaweb LTDA.
 * 	Todos os direitor reservados.
 *
 *  Está é uma API exemplo que facilita a utilização dos web services do Email Marketing.
 *
 * versao 1.0
 * mais detalhes em http://wiki.locaweb.com.br/pt-br/APIs_do_Email_Marketing
 */
using System;
using System.Collections.Generic;
using System.Web.Script.Serialization;
using Locaweb.Net;
using System.Net;

namespace Locaweb.EmailMarketingApi
{    
    public class EmailMkt
    {
        /// <summary>
        /// Nome do servidor
        /// </summary>
        private string hostname;
        /// <summary>
        /// Login usado no Email Marketing
        /// </summary>
        private string login;
        /// <summary>
        /// Chave gerada para uso dessa API
        /// </summary>
        private string chaveApi;

        private const string HOSTNAME_SUFIX = "locaweb.com.br";        

        public EmailMkt(string hostname, string login, string chaveApi)
        {
            this.hostname = hostname;
            this.login = login;
            this.chaveApi = chaveApi;
        }

        #region metodos publicos

        #region metodos de listagem de contatos

        /*
         * Os métodos de listagem possuem o parâmetro pagina. Ele informa qual página da pesquisa deve ser retornada.
         * Atualmente o limite de contatos por página é de 25mil contatos por página.
         * Por isso, caso tenha 40mil contatos em sua base por exemplo, precisará fazer 2 chamadas passando 
         * o parâmetro pagina=1 (que devolverá os contatos de 1 a 24999) e em seguida pagina=2 (que devolverá os contatos de 25000 a 40000) 
         */

        public List<Contato> retornaContatosValidos(int pagina)
        {
            return this.retornaContatosPorStatus(pagina, "validos");
        }
                
        public List<Contato> retornaContatosInvalidos(int pagina)
        {
            return this.retornaContatosPorStatus(pagina, "invalidos");
        }

        public List<Contato> retornaContatosNaoConfirmados(int pagina)
        {
            return this.retornaContatosPorStatus(pagina, "nao_confirmados");
        }

        public List<Contato> retornaContatosDescadastrados(int pagina)
        {
            return this.retornaContatosPorStatus(pagina, "descadastrados");
        }
        #endregion

        #endregion

        #region metodos privados

        /// <summary>
        /// Retorna todos os contatos. Se o número de contatos forsuperior a xxxx contatos, os contatos são quebrados em páginas de xxxx elementos.	 
        /// </summary>
        /// <param name="pagina">Número da página</param>
        /// <returns></returns>
        private List<Contato> retornaContatosPorStatus(int pagina, string status)
        {
            List<Contato> lcontatos = new List<Contato>();

            string urlApi = string.Format("http://{0}.{1}/admin/api/{2}/contacts/{3}?chave_api={4}&pagina={5}",
                                       this.hostname,
                                       HOSTNAME_SUFIX,
                                       this.login,
                                       status,
                                       this.chaveApi,
                                       pagina);

            string json = "";
            try
            {
                json = HttpClient.GET(urlApi);
            }
            catch (WebException e)
            {
                if (e.Response == null)
                {
                    Console.WriteLine("Ocorreu algum problema ao pegar a mensagem de erro");
                }

                using (WebResponse response = e.Response)
                {
                    HttpWebResponse httpResponse = (HttpWebResponse)response;

                    if (httpResponse.StatusCode == HttpStatusCode.NotFound)
                    {
                        return lcontatos;
                    }
                    else
                    {
                        throw e;
                    }
                }
            }
            
            JavaScriptSerializer serializer = new JavaScriptSerializer();            
            lcontatos = serializer.Deserialize<List<Contato>>(json);

            return lcontatos;
        }

        #endregion
    }
}

File: php/EmailMkt.php
<?php
/**
 *  Copyright (c) 2009, Locaweb LTDA.
 * 	Todos os direitor reservados.
 *
 *  Está é uma API exemplo que facilita a utilização dos web services do Email Marketing.
 *
 * @version 1.0
 * @see wiki
 */
class EmailMkt {

	private $HOSTNAME_SUFIX='tecnologia.ws';

	/**
	 * Nome do servidor.
	 */
    private $hostName;

    /**
     * Login usado no Email Marketing.
     */
	private $login;

	/**
	 * Chave gerada para uso dessa API.
	 */
	private $chaveApi;

	/**
	 * @param string hostName Login usado no Email Marketing.
	 * @param string Login usado no Email Marketing.
	 * @param string Chave gerada para uso dessa API.
	 */
	public function EmailMkt($hostName, $login, $chaveApi) {
		$this->hostName = $hostName;
		$this->login = $login;
		$this->chaveApi = $chaveApi;
	}

	public function retornaContatosValidos($pagina='1'){
		return $this->retornaContatosPorStatus($pagina,'validos');
	}

	public function retornaContatosInvalidos($pagina='1'){
		return $this->retornaContatosPorStatus($pagina,'invalidos');
	}

	public function retornaContatosNaoConfirmados($pagina='1'){
		return $this->retornaContatosPorStatus($pagina,'nao_confirmados');
	}

	public function retornaContatosDescadastrados($pagina='1'){
		return $this->retornaContatosPorStatus($pagina,'descadastrados');
	}

	/**
	 * Retorna todos os contatos. Se o número de contatos for
	 * superior a xxxx contatos, os contatos são quebrados em páginas
	 * de xxxx elementos.
	 *
	 * @param string pagina Número da página
	 * @return array Contatos ou null se não tiver contatos.
	 */
	private function retornaContatosPorStatus($pagina='1', $status) {

		$url = "http://{$this->hostName}.{$this->HOSTNAME_SUFIX}/admin" .
				"/api/{$this->login}/contatos/{$status}?chave_api={$this->chaveApi}&pagina={$pagina}";

		$resultado = $this->enviaRequisicao($url);
		if($resultado==null) {
			return null;
		}
		$resultado = json_decode($resultado);

		return $resultado;
	}

	private function enviaRequisicao($url) {
		$curl = curl_init();
		curl_setopt($curl, CURLOPT_URL, $url);
		curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
		$resultado_http = curl_exec($curl);
		$http_code = curl_getinfo($curl, CURLINFO_HTTP_CODE);
		curl_close($curl);

		if($http_code == '200'){
			return $resultado_http;
		}
		elseif ($http_code == '404') {
			return null;
		}
		elseif ($http_code == '500') {
			throw new Exception("Erro interno no servidor: $resultado_http");
		}
		elseif ($http_code == '401') {
			throw new Exception("Nao autorizado");
		}
		elseif ($http_code == '400') {
			throw new Exception("Parametros invalidos");
		}
		else{
			throw new Exception("Erro inesperado: statusCode:$http_code, mensagem:$resultado_http");
			return null;
		}
	}
}

?>
